File: backend/app/services/scrapers/ats_scrapers/lever_scraper.py
from datetime import datetime, timezone
from typing import Final, Optional

from app.schemas.scraped_job import ScrapedJob
from app.services.scrapers.ats_scrapers.base_ats_scraper import BaseATSScraper


class LeverScraper(BaseATSScraper):
    BASE_URL: Final[str] = "https://api.lever.co/v0/postings/"
    PARAMS: Final[dict] = {
        "mode" : "json"
        }

    def __init__(self):
        super().__init__(
            base_url = self.BASE_URL, 
            params = self.PARAMS
        )
    
    # 3. LEVER SCRAPER
    def map_to_scraped_job(self, job: dict, company_name: str) -> Optional[ScrapedJob]:
        title = job.get('text')
        url = job.get('hostedUrl')
        
        categories = job.get('categories') or {}
        loc_name = categories.get('location')

        if not title or not url or not loc_name:
            return None

        created_at = job.get('createdAt')
        iso_time = datetime.fromtimestamp(created_at / 1000.0, tz=timezone.utc).isoformat() if created_at else None

        return ScrapedJob(
            title=title,
            location=loc_name,
            posted_at=iso_time,
            url=url,
            company=company_name,
            platform="Lever"
        )

File: job-scraper-service/internal/scrapers/ats_scrapers/lever_scraper.go
package ats_scraper

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"job-scraper-service/internal/config"
	"job-scraper-service/internal/models"
	scraper "job-scraper-service/internal/scrapers"
)

type LeverScraper struct {
	scraper.BaseScraper
	BaseURL string
}

func NewLeverScraper(client *http.Client, cfg *config.ScraperConfig) *LeverScraper {
	return &LeverScraper{
		BaseScraper: scraper.NewBaseScraper(client, cfg, 250, 250, 250),
		BaseURL:     "https://api.lever.co/v0/postings/",
	}
}

type LeverJob struct {
	Text             string `json:"text"`
	HostedURL        string `json:"hostedUrl"`
	CreatedAt        int64  `json:"createdAt"`
	DescriptionPlain string `json:"descriptionPlain"`
	Categories       struct {
		Location string `json:"location"`
	} `json:"categories"`
}

func (l *LeverScraper) Fetch(ctx context.Context, company string, validationChan chan<- models.ScrapedJob) error {
	l.BaseScraper.Semaphore <- struct{}{}
	defer func() { <-l.BaseScraper.Semaphore }()

	if err := l.BaseScraper.RateLimiter.Wait(ctx); err != nil {
		return err
	}

	url := fmt.Sprintf("%s%s?mode=json", l.BaseURL, company)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}

	resp, err := l.BaseScraper.Client.Do(req)
	if err != nil {
		log.Printf("[Lever Fetch] Error for %s: %v", company, err)
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("lever api error status: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	// Lever returns a direct array of objects, not wrapped in a parent struct
	var rawJobs []LeverJob
	if err := json.Unmarshal(body, &rawJobs); err != nil {
		return err
	}

	for _, rawJob := range rawJobs {
		var postedAt *time.Time
		if rawJob.CreatedAt > 0 {
			// Convert Lever's Unix millisecond timestamp to time.Time
			parsedTime := time.UnixMilli(rawJob.CreatedAt)
			postedAt = &parsedTime
		}

		job := models.ScrapedJob{
			Title:       rawJob.Text,
			Company:     company,
			Platform:    models.Lever,
			Location:    rawJob.Categories.Location,
			Description: rawJob.DescriptionPlain,
			URL:         rawJob.HostedURL,
			PostedAt:    postedAt,
		}
		validationChan <- job
	}

	return nil
}

func (l *LeverScraper) Enrich(ctx context.Context, job models.ScrapedJob) (models.ScrapedJob, error) {
	return job, nil
}
